lib/puppet/configurer.rb

```diff
@@ -71,13 +73,18 @@ def retrieve_catalog(query_options)
 
         if result
           @cached_catalog_status = 'on_failure'
+
+          if result.environment == @environment
+            Puppet.info "Using cached catalog from environment '#{result.environment}'"
+          else
+            Puppet.err "Not using cached catalog because its environment '#{result.environment}' does not match '#{@environment}'"
+            result = nil
```

Comment:
I think we decided that the report shouldn't indicate the cached catalog was used ("on_failure") if we don't actually use the cache, so I think we need to only set it above if `result.environment == @environment`